### PR TITLE
feat: adding GitRepository resource support in flux

### LINKED ISSUE #18781: Adding support for `GitRepository` for `flux` manager
### What would you like Renovate to be able to do?

I would like to add support for updating `GitRepository` resources for the `flux` manager.
This will be helpful for me since a lot of my resources in the flux cluster are from a git repo.
This feature requests is taken from https://github.com/renovatebot/renovate/issues/12139#issuecomment-1230550840

### If you have any ideas on how this should be implemented, please tell us here.

- seperate the helm resources from git resources in the `ResourceFluxManifest` type
- add a switch case for `GitRepository` kind of resource in the flux `extract.ts`
- read and parse the content
- if a `commit` key is provided then the datasource is `GitRefDatasource` with the commit as the `currentDigest`
- if a `tag` key is provided then the datasource will be figured automatically from either `GithubTagsDatasource`,`GitlabTagsDatasource`, `BitBucketTagsDatasource`, or `GitTagsDatasource`
- dependancies are returned as usual

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by 5cat**
## Changes

`flux` manager now can consume `GitRepository` resources and update based `commit` or `tag` keys
## Context

Closes #18781
This is also related to https://github.com/renovatebot/renovate/issues/12139#issuecomment-1230550840

## Documentation (please check one with an [x])

- [X] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/flux/extract.ts
Comment (5cat) [reply lib/modules/manager/flux/extract.ts]: Mybad, forgot about it.

File: lib/modules/manager/flux/types.ts
Comment (viceice) [inline lib/modules/manager/flux/types.ts:59]: that's not the right way. this interface represent a yaml kubernetes manifest file, so it shouldn't users for other purpose.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/flux/types.ts
Comment (5cat) [reply lib/modules/manager/flux/types.ts:59]: Hey @viceice, I just learned TypeScript 10 hours ago so execuse me for not knowing.
But the interface was already not representing a single file before my changes.
Knowing that, do you wish me to work on isolating manifest objects and update the code logic accordingly?

Comment (viceice) [reply lib/modules/manager/flux/types.ts:59]: don't worry. was just a short review. will check the code tomorrow and make more suggestions then

Comment (5cat) [reply lib/modules/manager/flux/types.ts:59]: Hey @viceice , I think I know why. Because there can be multiple resources in the same yaml file, so that is why it was written that way. So the way I extended the code is okay.
Will mark this as resolved.

File: lib/modules/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/flux/readme.md:11]:
```suggestion
The `flux` manager only extracts `helm` dependencies for `HelmRelease` resources linked to `HelmRepository` or `GitRepository` sources.
```

File: lib/modules/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/flux/readme.md:14]:
```suggestion
Renovate does not support OCI `HelmRepository` sources, those with `type: oci`.
```

A list with one item is a bit silly. Let's rewrite into a single sentence. 😉

File: lib/modules/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/flux/readme.md:27]:
```suggestion
Renovate can update `git` references from `GitRepository` resources.
```

File: lib/modules/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/flux/readme.md:29]:
```suggestion
The `flux` manager only updates `GitRepository` fields that have a `tag` or `commit` key.
```

File: lib/modules/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/flux/readme.md:23]:
```suggestion
Renovate updates `HelmRelease` resources coming from `GitRepository` by updating the `GitRepository` resource.
```

We're using `resources` and `sourced` too much. I've tried to rewrite this sentence. I'm still not happy with it, but it's better than before.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some documentation improvements.

File: lib/modules/manager/flux/readme.md
Comment (5cat) [reply lib/modules/manager/flux/readme.md:21]: Thank you @HonkingGoose , Your sentences are much clearer I like them and happy with them.

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:155]: use for-of loops and a single deps result variable.

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:61]: we need to use a single releases variable, otherwise we get the deps in wrong order and the auto replace will fail.

so all deps need to stay in same order as they occured inside the file.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [reply lib/modules/manager/flux/extract.ts:61]: yes, sorry. I mean updatable dependencies like helm release or git repo.

**Review by rarkins — DISMISSED** [review]

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:166]: revert this and use resource directly, typescript is smart enough to narrow the type

image

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:190]: same here

**Comment by 5cat** [comment]
I dont know why the re-request features dismiss the other reviewers, forgive me for this.

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:158]: this should be moved one level up, so we don't do that loop for every manifest again.

Comment (5cat) [reply lib/modules/manager/flux/extract.ts:158]: Do you mean that I will remove the `context` variable and replace it with `helmrepositories` then create the `helmrepositories` array in the `extractAllPackageFiles` function and pass it down via the parameters `extractAllPackageFiles->resolveManifest->resolveResourceManifest`?

Because the loop for the manifests happen in `extractAllPackageFiles`.
And the `resolveManifest` does handle both system and resource manifests, so passing down `helm` resources specifically from there while system resolve might not need them might be an anti-pattern, although i dont know the convention of typescript so im not sure.

What do you think?

**Comment by viceice** [comment]
> I dont know why the re-request features dismiss the other reviewers, forgive me for this.

the git push has dismissed the review 😉

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [reply lib/modules/manager/flux/extract.ts:158]: yes, but inline `resolveManifest` to `extractAllPackageFiles`, so we don't need that small extra function

**Review by viceice — DISMISSED** [review]

**Comment by viceice** [comment]
needs deconfliction

**Comment by 5cat** [comment]
@viceice the main branch does not pass the tests 

```
/usr/src/app/lib/modules/manager/flux/extract.spec.ts
  1:27  error  Unable to resolve path to module 'common-tags'  import/no-unresolved

/usr/src/app/lib/modules/manager/gomod/artifacts.spec.ts
  1:27  error  Unable to resolve path to module 'common-tags'  import/no-unresolved

/usr/src/app/lib/modules/manager/hermit/update.spec.ts
   1:27  error  Unable to resolve path to module 'common-tags'     import/no-unresolved
  12:27  error  Invalid type "any" of template literal expression  @typescript-eslint/restrict-template-expressions
  22:27  error  Invalid type "any" of template literal expression  @typescript-eslint/restrict-template-expressions

/usr/src/app/lib/modules/manager/mint/extract.spec.ts
  1:27  error  Unable to resolve path to module 'common-tags'  import/no-unresolved

/usr/src/app/lib/modules/manager/terraform/extract.spec.ts
  1:27  error  Unable to resolve path to module 'common-tags'  import/no-unresolved

/usr/src/app/lib/modules/manager/tflint-plugin/extract.spec.ts
  1:27  error  Unable to resolve path to module 'common-tags'  import/no-unresolved

/usr/src/app/lib/util/git/errors.spec.ts
  1:27  error  Unable to resolve path to module 'common-tags'  import/no-unresolved

/usr/src/app/lib/util/http/github.spec.ts
  1:27  error  Unable to resolve path to module 'common-tags'  import/no-unresolved

/usr/src/app/lib/util/markdown.spec.ts
   1:27  error  Unable to resolve path to module 'common-tags'                                          import/no-unresolved
  21:7   error  Operands of '+' operation with any is possible only with string, number, bigint or any  @typescript-eslint/restrict-plus-operands

✖ 12 problems (12 errors, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "eslint-fix" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
I fixed the conflicts locally but these errors have been inherited from main to my branch when I merged main into it. Should I push my code?

**Comment by viceice** [comment]
@5cat you need to run yarn install after merge, main is ok

https://github.com/renovatebot/renovate/actions/runs/3468750253

**Comment by onedr0p** [comment]
@viceice any chance of another review or merge if all good? I have expanded on these changes for `OCIRepository` and `HelmRepository` OCI. I don't mind waiting just poking a bit, thanks!

File: lib/modules/manager/flux/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.spec.ts:246]: please prefer to match full result object with toEqual, so things can't silently break

Comment (5cat) [reply lib/modules/manager/flux/extract.spec.ts:246]: Hi @viceice nice to see you back : )
I tried to change every `toBe` to `toEqual` and I got this error.
```
 > podman run -it --env-file .env --rm -v "$PWD":/usr/src/app -w /usr/src/app renovatebot_local yarn test
yarn run v1.22.19
$ run-s generate:*
$ node tools/generate-imports.mjs
generating imports
> data/azure-pipelines-tasks.json
> data/debian-distro-info.json
> data/kubernetes-api.json5
> data/node-js-schedule.json
> data/ubuntu-distro-info.json
> node_modules/emojibase-data/en/shortcodes/github.json
generating hashes
$ run-s lint test-schema type-check strict-check jest
$ run-s ls-lint eslint prettier markdown-lint git-check doc-fence-check
$ ls-lint
$ eslint . --cache --report-unused-disable-directives

/usr/src/app/lib/modules/manager/flux/extract.spec.ts
   88:42  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  111:42  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  158:42  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  178:42  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  199:42  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  214:42  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  230:42  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  246:42  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  264:45  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  265:45  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  284:44  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  303:44  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  322:44  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  341:44  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
  360:44  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be

✖ 15 problems (15 errors, 0 warnings)
  15 errors and 0 warnings potentially fixable with the `--fix` option.

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "eslint" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "lint" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
So I applied the `lint-fix` and only some of the `toBe` have remained, and the line you mentioned is not one of them. 
```
  246:42  error  Use `toBe` when expecting primitive literals  jest/prefer-to-be
```
I git pulled and did `yarn install` so im kind of up to date with the main branch, did i miss a step that made the linter complain about this line?

Comment (viceice) [reply lib/modules/manager/flux/extract.spec.ts:246]: you should rewrite text like:

`expect(result).toEqual({ ... });`

so we've a single expectation.

Comment (onedr0p) [reply lib/modules/manager/flux/extract.spec.ts:246]: I updated my working branch with

```js
      expect(result?.deps[0]).toEqual({
        depName: 'renovate-repo',
        skipReason: 'unversioned-reference',
      });
```

but it seems like we need to target `result` in the expect directly

**Review by viceice — APPROVED** [review]

**Comment by 5cat** [comment]
Thank you @viceice @rarkins and @HonkingGoose For your help in this pull request. 
I always thought about a tool that auto-update my dependencies that is not hard dependent on github and was almost about to write my own until I saw this project and instead wanted to contribute to it with the features I needed.
You are awesome thank you for your continues support.

**Comment by HonkingGoose** [comment]
Thank you for the kind words! I'm happy to improve the docs! 🙂

**Pull request merged**
